cpp/test/sg/decisiontree_batchedlevel_algo.cu

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) 2019-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <decisiontree/quantile/quantile.h>
+#include <gtest/gtest.h>
+#include <raft/linalg/cublas_wrappers.h>
+#include <test_utils.h>
+#include <common/iota.cuh>
+#include <decisiontree/batched-levelalgo/builder.cuh>
+#include <memory>
+#include <raft/cuda_utils.cuh>
+#include <raft/handle.hpp>
+#include <random/make_blobs.cuh>
+#include <random/make_regression.cuh>
+
+namespace ML {
+namespace DT {
+
+struct DtTestParams {
+  int M, N, nclasses, max_depth, nbins;
+  float min_gain;
+  CRITERION splitType;
+  unsigned long long seed;
+};
+
+::std::ostream& operator<<(::std::ostream& os, const DtTestParams& dims) { return os; }
+
+template <typename T, typename L, typename I = int>
+class DtBaseTest : public ::testing::TestWithParam<DtTestParams> {
+ protected:
+  void SetUp()
+  {
+    inparams = ::testing::TestWithParam<DtTestParams>::GetParam();
+    handle.reset(new raft::handle_t);
+    CUDA_CHECK(cudaStreamCreate(&stream));
+    handle->set_stream(stream);
+    set_tree_params(params,
+                    inparams.max_depth,
+                    1 << inparams.max_depth,
+                    1.f,
+                    inparams.nbins,
+                    0,
+                    inparams.nbins,
+                    inparams.min_gain,
+                    inparams.splitType,
+                    128);
+    data   = std::make_unique<rmm::device_uvector<T>>(inparams.M * inparams.N, stream);
+    labels = std::make_unique<rmm::device_uvector<L>>(inparams.M, stream);
+    tmp    = rmm::device_uvector<T>(inparams.M * inparams.N, stream);
+    prepareDataset(tmp.data());
+    auto alpha = T(1.0), beta = T(0.0);
```

Comment:
Let's split these into multiple lines for clarity and to avoid weird edge cases with multiple definitions on a single line.